### PR TITLE
chore: add member, mappable member and attribute caching

### PR DESCRIPTION
**Pull request by TimothyMakkison**
# Add member, mappable member and attribute caching

## Description
Added member, mappable member and attribute caching to improve performance and reduce memory usage.

Splitting up changes from #530 

## Benchmarks
10ms time and 3000KB memory savings
### Cache
|       Method |      Mean |     Error |    StdDev |      Gen0 |     Gen1 |  Allocated |
|------------- |----------:|----------:|----------:|----------:|---------:|-----------:|
|      Compile |  2.491 ms | 0.0391 ms | 0.0366 ms |  156.2500 |  31.2500 |  284.15 KB |
| LargeCompile | 64.276 ms | 1.2297 ms | 1.2628 ms | 1714.2857 | 714.2857 | 10713.9 KB |

### Original
|       Method |      Mean |     Error |    StdDev |      Gen0 |     Gen1 |   Allocated |
|------------- |----------:|----------:|----------:|----------:|---------:|------------:|
|      Compile |  2.640 ms | 0.0524 ms | 0.0490 ms |  179.6875 |  46.8750 |   320.39 KB |
| LargeCompile | 75.350 ms | 1.4524 ms | 1.3586 ms | 2333.3333 | 833.3333 | 14283.01 KB |
|       Method |      Mean |     Error |    StdDev |      Gen0 |     Gen1 |   Allocated |
|------------- |----------:|----------:|----------:|----------:|---------:|------------:|
|      Compile |  2.627 ms | 0.0428 ms | 0.0379 ms |  179.6875 |  39.0625 |   326.46 KB |
| LargeCompile | 72.979 ms | 1.1195 ms | 0.8740 ms | 2285.7143 | 571.4286 | 14280.24 KB |

## Checklist

- [x] The existing code style is followed
- [x] The commit message follows our guidelines
- [x] Performed a self-review of my code
- [x] Hard-to-understand areas of my code are commented
- [x] The documentation is updated (as applicable)
- [x] Unit tests are added/updated
- [x] Integration tests are added/updated (as applicable, especially if feature/bug depends on roslyn or framework version in use)

### CODE REVIEW

File: src/Riok.Mapperly/Helpers/SymbolExtensions.cs [hunk 1]
```diff
@@ -14,8 +15,8 @@ internal static class SymbolExtensions
         typeof(Version).FullName
     );
 
-    internal static bool HasAttribute(this ISymbol symbol, INamedTypeSymbol attributeSymbol) =>
-        symbol.GetAttributes().Any(a => SymbolEqualityComparer.Default.Equals(a.AttributeClass, attributeSymbol));
+    internal static bool HasAttribute(this ISymbol symbol, INamedTypeSymbol attributeSymbol, WellKnownTypes types) =>
```
Comment (latonz) [inline src/Riok.Mapperly/Helpers/SymbolExtensions.cs:18]: If this is moved to the `WellKnownTypes` class, it could be simplified to `bool HasAttribute<TAttribute>(ISymbol symbol) where TAttribute : Attribute`

File: src/Riok.Mapperly/Helpers/SymbolExtensions.cs [hunk 2]
```diff
@@ -56,46 +57,25 @@ internal static bool TryGetEnumUnderlyingType(this ITypeSymbol t, [NotNullWhen(t
...
 
-    internal static IEnumerable<IMethodSymbol> GetAllMethods(this ITypeSymbol symbol, string name) =>
-        symbol.GetAllMembers(name).OfType<IMethodSymbol>();
+    internal static IEnumerable<IMethodSymbol> GetAllMethods(this ITypeSymbol symbol, string name, WellKnownTypes types) =>
+        symbol.GetAllMembers(name, types).OfType<IMethodSymbol>();
 
-    internal static IEnumerable<IPropertySymbol> GetAllProperties(this ITypeSymbol symbol, string name) =>
-        symbol.GetAllMembers(name).OfType<IPropertySymbol>();
+    internal static IEnumerable<IPropertySymbol> GetAllProperties(this ITypeSymbol symbol, string name, WellKnownTypes types) =>
+        symbol.GetAllMembers(name, types).OfType<IPropertySymbol>();
```
Comment (latonz) [inline src/Riok.Mapperly/Helpers/SymbolExtensions.cs:67]: I'd move all modified methods of this class and `GetFields` to the `WellKnownTypes` or a new class `SymbolAccessor` and use the symbol as parameter.

File: src/Riok.Mapperly/Descriptors/WellKnownTypes.cs [hunk 3]
```diff
@@ -1,12 +1,17 @@
...
 
 namespace Riok.Mapperly.Descriptors;
 
 public class WellKnownTypes
 {
     private readonly Compilation _compilation;
     private readonly Dictionary<string, INamedTypeSymbol?> _cachedTypes = new();
+    private readonly Dictionary<ISymbol, ImmutableArray<AttributeData>> _cachedAttributes = new(SymbolEqualityComparer.Default);
+    private readonly Dictionary<ITypeSymbol, ISymbol[]> _allMembers = new(SymbolEqualityComparer.Default);
+    private readonly Dictionary<ITypeSymbol, IMappableMember[]> _allAccessibleMembers = new(SymbolEqualityComparer.Default);
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/WellKnownTypes.cs:14]: I'd use `ReadOnlyCollection` as value type of the dictionary, since these should never be modified.

File: src/Riok.Mapperly/Descriptors/WellKnownTypes.cs [hunk 4]
```diff
@@ -41,4 +46,66 @@ internal WellKnownTypes(Compilation compilation)
 
         return typeSymbol;
     }
+
+    public ImmutableArray<AttributeData> GetAttributes(ISymbol symbol)
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/WellKnownTypes.cs:50]: I think this member/attribute resolution doesn't really fit to `WellKnownTypes`. I'd create a new class named `SymbolAccessor` or similar which has a private field `WellKnownTypes`.

File: src/Riok.Mapperly/Descriptors/WellKnownTypes.cs [hunk 5]
```diff
@@ -41,4 +46,66 @@ internal WellKnownTypes(Compilation compilation)
...
+    }
+
+    private IMappableMember[] GetAllAccessibleMappableMembersCore(ITypeSymbol symbol)
+    {
+        return GetAllMembers(symbol)
+            .Where(x => !x.IsStatic && x.IsAccessible() && x.Kind is SymbolKind.Property or SymbolKind.Field)
+            .DistinctBy(x => x.Name)
+            .Select(MappableMember.Create)
+            .WhereNotNull()
+            .ToArray();
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/WellKnownTypes.cs:109]: I wouldn't collect to array here to match the signature of `GetAllMembersCore`.

**Review by latonz — CHANGES_REQUESTED** [review]

File: src/Riok.Mapperly/Descriptors/WellKnownTypes.cs (hunk 4 above)
Comment (TimothyMakkison) [reply src/Riok.Mapperly/Descriptors/WellKnownTypes.cs:50]: Should I move `GetAllMembers`, `GetAllAccessibleMappableMembers` and related dictionaries to this class?

**Comment by TimothyMakkison** [comment]
Thanks for the feedback, added `SymbolAccessor` etc.
Ran a quick check and memory usage is now 292.5 KB & 10715.43 KB not 1005 sure why

File: src/Riok.Mapperly/Descriptors/SymbolAccessor.cs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+using Microsoft.CodeAnalysis;
+using Riok.Mapperly.Helpers;
+using Riok.Mapperly.Symbols;
+
+namespace Riok.Mapperly.Descriptors;
+
+public class SymbolAccessor
+{
+    private readonly WellKnownTypes _types;
+    private readonly Dictionary<ISymbol, ImmutableArray<AttributeData>> _cachedAttributes = new(SymbolEqualityComparer.Default);
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/SymbolAccessor.cs:11]: I'd remove the `_cached` prefix, as the other cache fields don't have it either.

File: src/Riok.Mapperly/Descriptors/SymbolAccessor.cs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    internal IEnumerable<IFieldSymbol> GetAllFields(ITypeSymbol symbol) => GetAllMembers(symbol).OfType<IFieldSymbol>();
+
+    internal IReadOnlyCollection<ISymbol> GetAllMembers(ITypeSymbol symbol)
+    {
+        if (_allMembers.TryGetValue(symbol, out var members))
+        {
+            return members;
+        }
+
+        members = GetAllMembersCore(symbol).ToArray();
```
Comment (latonz) [inline src/Riok.Mapperly/Descriptors/SymbolAccessor.cs:58]: Do we really need to collect to array instead of to list? Wouldn't to list be faster here? Same for the accessible mappable members.

Comment (TimothyMakkison) [reply src/Riok.Mapperly/Descriptors/SymbolAccessor.cs:58]: I'm not sure. Why do you think List is faster?

Comment (TimothyMakkison) [reply src/Riok.Mapperly/Descriptors/SymbolAccessor.cs:58]: It looks like `members = GetAllMembersCore(symbol)` actually returns `ICollection<ISymbol>`. `ToArray` and `ToList` both have a check for `ICollection` using `CopyTo` under the hood. 
TLDR both should have the same performance (`ToArray` might is faster for an empty ICollection)

Comment (latonz) [reply src/Riok.Mapperly/Descriptors/SymbolAccessor.cs:58]: I think the current solution is fine for now (good performance while still keeping good readability and maintainability).

File: src/Riok.Mapperly/Descriptors/SymbolAccessor.cs [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    private readonly Dictionary<ITypeSymbol, IReadOnlyCollection<ISymbol>> _allMembers = new(SymbolEqualityComparer.Default);
+    private readonly Dictionary<ITypeSymbol, IReadOnlyCollection<IMappableMember>> _allAccessibleMembers =
+        new(SymbolEqualityComparer.Default);
+
+    public SymbolAccessor(WellKnownTypes types)
+    {
+        _types = types;
+    }
+
+    internal ImmutableArray<AttributeData> GetAttributes(ISymbol symbol)
```
[the remark this answers is not in the document]
Comment (TimothyMakkison) [reply src/Riok.Mapperly/Descriptors/SymbolAccessor.cs]: This initially prevented mapperly from working.

`AttributeDataAccessor` uses `Namespace` + `Name` to get the types name for lookup whereas `WellKnownTypes.Get` used `FullName`.

I've changed `WellKnownTypes` to interpolation. Hopefully this doesn't break anything.
```C#
var dataType = typeof(TData);
var attrSymbol = _types.Get($"{attrType.Namespace}.{attrType.Name}");

// vs
public INamedTypeSymbol Get(Type type) =>
        Get(type.FullName ?? throw new InvalidOperationException("Could not get name of type " + type));
```

**Comment by TimothyMakkison** [comment]
Thanks for the help. I've resolved the issues 😂

**Benchmark**
|       Method |      Mean |     Error |    StdDev |      Gen0 |     Gen1 |   Allocated |
|------------- |----------:|----------:|----------:|----------:|---------:|------------:|
|      Compile |  2.367 ms | 0.0454 ms | 0.0606 ms |  152.3438 |  35.1563 |   273.75 KB |
| LargeCompile | 61.346 ms | 1.1614 ms | 1.1407 ms | 1714.2857 | 571.4286 | 10581.14 KB |

**Review by latonz — APPROVED** [review]
